File: src/publisher.cpp
#include "ros/ros.h"
#include "performance_test/SuperAwesome.h"


int main(int argc, char **argv) {
  ros::init(argc, argv, "cpp_publisher");
  ros::NodeHandle n;

  ros::Publisher publisher = n.advertise<performance_test::SuperAwesome>("performance_test", 10);

  ros::Rate loop_rate(10);

  while (ros::ok()) {
    performance_test::SuperAwesome msg;
    msg.content = "Content sample";
    ROS_INFO("%s", msg.content.c_str());

    publisher.publish(msg);

    ros::spinOnce();

    loop_rate.sleep();
  }
  return 0;
} 

File: scripts/publisher.py
#!/usr/bin/env python

import rospy
from performance_test.msg import SuperAwesome

def publisher():
    """
	Publishes SuperAwesome message using rospy.
	"""
    # Node, handler and publisher init
    rospy.init_node("python_publisher", anonymous=True)

    pub = rospy.Publisher("performance_test", SuperAwesome, queue_size = 100000)
    
    rate_param = rospy.get_param("/publish_rate")
    rate = rospy.Rate(rate_param)

    while not rospy.is_shutdown():
        msg = SuperAwesome()
        # Since SuperAwesome is only allowed to contain string, use to send time at message creation
        msg.content = str(rospy.Time.now())
        pub.publish(msg)

        # Check latest rate param and update if different
        if(rate_param != rospy.get_param("/publish_rate")):
            rate_param = rospy.get_param("/publish_rate")
            rate = rospy.Rate(rate_param)

        # Sleep thread until next loop
        rate.sleep()

if __name__ == "__main__":
    try:
        publisher()
    except rospy.ROSInterruptException:
        pass 
